AdminServer/appscale/admin/__init__.py

```diff
@@ -0,0 +1,400 @@
+""" A server that handles application deployments. """
+
+import argparse
+import logging
+import socket
+import sys
+import time
+
+from appscale.common import appscale_info
+from appscale.common.constants import HTTPCodes
+from appscale.common.constants import LOG_FORMAT
+from appscale.common.ua_client import UAClient
+from appscale.common.ua_client import UAException
+from appscale.common.unpackaged import APPSCALE_PYTHON_APPSERVER
+from tornado import gen
+from tornado.options import options
+from tornado import web
+from tornado.escape import json_decode
+from tornado.escape import json_encode
+from tornado.ioloop import IOLoop
+from . import utils
+from . import constants
+from .constants import (
+  CustomHTTPError,
+  MAX_DEPLOY_TIME,
+  OperationTimeout,
+  REDEPLOY_WAIT,
+  VALID_RUNTIMES
+)
+from .operation import CreateVersionOperation
+from .operations_cache import OperationsCache
+
+sys.path.append(APPSCALE_PYTHON_APPSERVER)
+from google.appengine.api.appcontroller_client import AppControllerException
+
+
+# The state of each operation.
+operations = OperationsCache()
+
+
+@gen.coroutine
+def wait_for_port_assignment(operation_id, deadline, acc):
+  """ Waits until port is assigned for version.
+
+  Args:
+    operation_id: A string specifying an operation ID.
+    deadline: A float containing a unix timestamp.
+    acc: An AppControllerClient.
+  Raises:
+    OperationTimeout if the deadline is exceeded.
+  """
+  try:
+    operation = operations[operation_id]
+  except KeyError:
+    raise OperationTimeout('Operation no longer in cache')
+
+  project_id = operation.project_id
+
+  while True:
+    if time.time() > deadline:
+      message = 'Deploy operation took too long.'
+      operation.set_error(message)
+      raise OperationTimeout(message)
+
+    try:
+      info_map = acc.get_app_info_map()
+    except AppControllerException as error:
+      logging.warning('Unable to fetch info map: {}'.format(error))
+      yield gen.sleep(1)
+      continue
+
+    try:
+      raise gen.Return(info_map[project_id]['nginx'])
```

Comment:
Python 2 does not allow you to return values from a generator. This is kind of a hack required by tornado for coroutines. http://www.tornadoweb.org/en/stable/gen.html#tornado.gen.coroutine